### PR TITLE
Transform remove

### PR DESCRIPTION
**Pull request by bloderxd**
This PR proposes the implementation of a Transform that removes elements from the AST.
This kind of Transform has three possible apis:

1. Remove the entire element

```kotlin
// All functions called "helloWorld" will be removed from the AST
namedFunction({ name == "helloWorld" }) { c ->
  Transform.remove(remove = c)
}
```

2.  Remove a single element from an element

```kotlin
/**
 * All "println("")" expressions that are inside a function called "helloWorld" will be removed from the 
 * AST
**/
namedFunction({ name == "helloWorld" }) { c ->
  Transform.remove(
    removeIn = c,
    declaration = """ println("") """.expressionIn(c)
  )
}
```

3. Remove multiple elements from an element

```kotlin
/**
 * All "println("")" and "println("test")" expressions that are inside a function called "helloWorld" will 
 * be removed from the AST
**/
namedFunction({ name == "helloWorld" }) { c ->
  Transform.remove(
    removeIn = c,
    declarations = listOf(
      """ println("") """.expressionIn(c),
      """ println("test") """.expressionIn(c)
    )
  )
}
```

This is related to https://github.com/arrow-kt/arrow-meta/issues/177

### CODE REVIEW

File: compiler-plugin/src/main/kotlin/arrow/meta/quotes/Quote.kt [hunk 1]
```diff
@@ -267,29 +268,43 @@ fun <K : KtElement> KtFile.sourceWithTransformationsAst(mutations: ArrayList<Tra
...
-      }
+      is Transform.Replace -> dummyFile = transform.transform(dummyFile)
+      is Transform.Remove -> dummyFile = transform.transform(dummyFile)
       Transform.Empty -> Unit
     }
   }
   return Writer.write(dummyFile)
 }
 
+private fun <T : KtElement> Transform.Replace<T>.transform(file: Node.File): Node.File = MutableVisitor.preVisit(file) { element, _ ->
```
Comment (raulraja) [inline compiler-plugin/src/main/kotlin/arrow/meta/quotes/Quote.kt]:
```suggestion
private fun <T : KtElement> Transform.Replace<T>.replace(file: Node.File): Node.File = MutableVisitor.preVisit(file) { element, _ ->
```

File: compiler-plugin/src/main/kotlin/arrow/meta/quotes/Quote.kt [hunk 2]
```diff
@@ -267,29 +268,43 @@ fun <K : KtElement> KtFile.sourceWithTransformationsAst(mutations: ArrayList<Tra
...
+    if (element != null && element == replacing.ast) {
+        val newContents = newDeclarations.joinToString("\n") { it.value?.text ?: "" }
+        println("Replacing ${element.javaClass} with ${newDeclarations.map { it.value?.javaClass }}: newContents: \n$newContents")
+        element.dynamic = newContents
+        element
+    } else element
+}
+
+private fun <T : KtElement> Transform.Remove<T>.transform(file: Node.File): Node.File {
```
Comment (raulraja) [inline compiler-plugin/src/main/kotlin/arrow/meta/quotes/Quote.kt]:
```suggestion
private fun <T : KtElement> Transform.Remove<T>.remove(file: Node.File): Node.File {
```

File: compiler-plugin/src/main/kotlin/arrow/meta/quotes/Quote.kt [hunk 3]
```diff
@@ -267,29 +268,43 @@ fun <K : KtElement> KtFile.sourceWithTransformationsAst(mutations: ArrayList<Tra
...
-          if (element != null && element == replacingNode) {
-            val newContents = transform.newDeclarations.joinToString("\n") { it.value?.text ?: "" }
-            println("Replacing ${element.javaClass} with ${transform.newDeclarations.map { it.value?.javaClass }}: newContents: \n$newContents")
-            element.dynamic = newContents
-            element
-          } else element
-        }
-        Unit
-      }
+      is Transform.Replace -> dummyFile = transform.transform(dummyFile)
```
Comment (raulraja) [inline compiler-plugin/src/main/kotlin/arrow/meta/quotes/Quote.kt]:
```suggestion
      is Transform.Replace -> dummyFile = transform.replace(dummyFile)
```

File: compiler-plugin/src/main/kotlin/arrow/meta/quotes/Quote.kt [hunk 4]
```diff
@@ -267,29 +268,43 @@ fun <K : KtElement> KtFile.sourceWithTransformationsAst(mutations: ArrayList<Tra
...
+      is Transform.Remove -> dummyFile = transform.transform(dummyFile)
```
Comment (raulraja) [inline compiler-plugin/src/main/kotlin/arrow/meta/quotes/Quote.kt]:
```suggestion
      is Transform.Remove -> dummyFile = transform.remove(dummyFile)
```

**Comment by raulraja** [comment]
We need to also add a table for the Transforms inside the Quote menu that follows the same structure as the quotes table so users have visibility into what transforms are possible. We have an issue describing the ones we know we want to achieve at the moment in #177 . /cc @ahinchman1, @franciscodr can you add an issue to have this for the KC milestone

File: testing-plugin/src/test/kotlin/arrow/meta/plugin/testing/plugins/transform/TransformRemovePlugin.kt [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+      meta(
+        namedFunction({ name == "transformRemove" }) { f ->
+            Transform.remove(f)
+        }
+      )
+    }
+
+private val Meta.transformRemoveSingleElementFromContext: Plugin
+  get() =
+    "Transform Remove" {
```
Comment (ahinchman1) [inline testing-plugin/src/test/kotlin/arrow/meta/plugin/testing/plugins/transform/TransformRemovePlugin.kt]: Let's name these "Remove Single Element From Context" and so on for more descriptive testing!

File: compiler-plugin/src/main/kotlin/arrow/meta/quotes/Quote.kt [hunk 6]
```diff
@@ -267,29 +268,43 @@ fun <K : KtElement> KtFile.sourceWithTransformationsAst(mutations: ArrayList<Tra
...
-            val newContents = transform.newDeclarations.joinToString("\n") { it.value?.text ?: "" }
-            println("Replacing ${element.javaClass} with ${transform.newDeclarations.map { it.value?.javaClass }}: newContents: \n$newContents")
-            element.dynamic = newContents
-            element
-          } else element
-        }
-        Unit
-      }
+      is Transform.Replace -> dummyFile = transform.replace(dummyFile)
+      is Transform.Remove -> dummyFile = transform.remove(dummyFile)
```
Comment (ahinchman1) [inline compiler-plugin/src/main/kotlin/arrow/meta/quotes/Quote.kt]: Nice work! I'm glad you moved the mechanisms into their own functions.

**Pull request merged**